### PR TITLE
#4067 Initial design of the new plugins - pre-post backup and restore

### PR DESCRIPTION
**Pull request by brito-rafa**
Signed-off-by: Rafael Brito <rbrito@vmware.com>

Thank you for contributing to Velero!

# Please add a summary of your change
Velero should provide a way to trigger actions before and after each backup and restore.
**Important**: These proposed plugin hooks are fundamentally different from the existing plugin hooks, BackupItemAction and RestoreItemAction, which are triggered per item during backup and restore, respectively.
The proposed plugin hooks are to be executed only once.
# Does your change fix a particular issue?

Fixes #(issue)
#4067 

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required`.
- [ NA] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Comment by dsu-igeek** [comment]
Let's review individually and discuss in the next community meeting before merging.

File: design/new-prepost-backuprestore-plugin-hooks.md
Comment (sseago) [inline design/new-prepost-backuprestore-plugin-hooks.md:56]: Does this mean that if the PreBackupAction modifies/creates resources in the cluster, those new resources and/or changes to resources will be reflected in the backup? On the other hand, if resources are pulled from the cluster for backup after running the pre hook, does that mean Deployments, etc. will be backed up in quiesced state if such a plugin is registered (and thus, on restore, replicas will still be 0)?

File: design/new-prepost-backuprestore-plugin-hooks.md
Comment (brito-rafa) [reply design/new-prepost-backuprestore-plugin-hooks.md:34]: Agree.

File: design/new-prepost-backuprestore-plugin-hooks.md
Comment (brito-rafa) [reply design/new-prepost-backuprestore-plugin-hooks.md:56]: yes.

File: design/new-prepost-backuprestore-plugin-hooks.md
Comment (brito-rafa) [reply design/new-prepost-backuprestore-plugin-hooks.md:34]: Just a note from what we discussed during the 2021-09-14 community meeting: it is better for the postback plugins to run after the upload of the backup on the BSL. Why? We want to be positive the backup is really completed before doing any further operation. We want to have separate backup/restore phases for these plugins. Let's have their logs persisted separately as well. On the pre restore actions, run proactive sync on the BSL to make sure the latest backup is there.

Comment (brito-rafa) [reply design/new-prepost-backuprestore-plugin-hooks.md:34]: Added the changes on the design as per feedback from the last call. Will go for another round of feedback.

**Comment by jaidevmane** [comment]
Would love to participate in discussions on this. Our team has a use case for this workflow, where currently we are asking users to run through a two step restore process, and using a post-restore/backup hook would enable us to abstract away the step for them.
https://github.com/vmware-tanzu/velero/issues/1981

**Comment by brito-rafa** [comment]
Please take a look here at this code: this is a functional implementation of this design, without the plugins logging - https://github.com/cschaefer-vmw/velero

For the design, the only thing missing is defining how the logging of post-backup and post-restore plugins will act behind the scenes (the plan is expanding existent interfaces).

File: design/new-prepost-backuprestore-plugin-hooks.md
Comment (reasonerjt) [inline design/new-prepost-backuprestore-plugin-hooks.md:137]: A couple of questions:
1.  If a PostBackup action fails, do we consider the backup successful and usable?
2. How does user get the execution log of PostBackup plugin?
3. Once registered, will the plugins per backup/restore take effect to all backups and restores?  Will there be a mechanism such these plugins can be executed selectively?
4. How does user define timeout?  And how will it be handled?

Comment (brito-rafa) [reply design/new-prepost-backuprestore-plugin-hooks.md:137]: Answering:
1. Yes
2. The plan is to implement an interface to persist the post backup log on the same location as the backup log (not necessarily inside the gzip file, but on the same directory as a different name).
3. Yes. But the plugin developer can implement on their plugin a mechanism so the plugin to execute selectively, for example, a config map. 
4. This design does not touch on timeouts since current hooks do not provide any provisioning for such. There is an en-route design about adding timeout on plugins, which I propose to bring back to the discussion table once we have these new hooks implemented: https://github.com/vmware-tanzu/velero/pull/3355 .

Comment (reasonerjt) [reply design/new-prepost-backuprestore-plugin-hooks.md:137]: > The plan is to implement an interface to persist the post backup log on the same location as the backup log (not necessarily inside the gzip file, but on the same directory as a different name).

If we provide a way for users to collect the backup log, we should also provide a way for users to see the log of the plugin executions.

For 3, if we agree that for most users the plugins should not always be triggered for different backup and restores we should provide a way in the framework to enable user to choose what plugins to be executed or skipped.  IMO, Iinstead of relying on users to workaround, it should be covered in the design.

Comment (brito-rafa) [reply design/new-prepost-backuprestore-plugin-hooks.md:137]: On point #4, when we create the post backup log, we will make sure the `velero` command line will be able to collect them. For point #3, do we have similar functionality on existent hooks + plugins? If not, I am glad to put some design doc after this one goes through.

Comment (reasonerjt) [reply design/new-prepost-backuprestore-plugin-hooks.md:137]: @brito-rafa 

> On point #4, when we create the post backup log, we will make sure the velero command line will be able to collect them. 

I think you mean point 2, it's probably OK to put it in another PR to make velero CLI return _additional logs_ as it's also needed for itemSnapshotter plugins.

But for point 3, the existing hooks are at a resource level, which doesn't impact all backups and restores, additionally, the user who creates the backup can edit the annotations to control whether the hooks will be executed.  We need to address that for the pre/post backup/restore plugin before this feature is delivered.  Therefore, ideally, it should be covered in this design.

Comment (brito-rafa) [reply design/new-prepost-backuprestore-plugin-hooks.md:137]: I concur on annotations. I added a section on the design document about this.

File: design/new-prepost-backuprestore-plugin-hooks.md
Comment (reasonerjt) [inline design/new-prepost-backuprestore-plugin-hooks.md:693]: Thanks for adding this part @brito-rafa 

I think this approach works but I wanna confirm with other maintainers about the usability in the e2e flow:

1.  When a user uses CLI to create backup how does he/she set the annotations?
2. How does the user know the name of the plugins?

Comment (brito-rafa) [reply design/new-prepost-backuprestore-plugin-hooks.md:693]: I am ok (and volunteer) to add capabilities on the client-side to manage plugins execution on each backup or restore (i.e., adding a flag to skip plugins, which translates in adding annotations on the object). I would say this change would go along with the CLI change to download the plugin logs, which we discussed can be implemented incrementally. The user can run `velero plugin get` and use the same name from the plugin list to cherry-pick the plugin to be skipped.

**Review by reasonerjt — APPROVED** [review]
LGTM

File: design/new-prepost-backuprestore-plugin-hooks.md
Comment (brito-rafa) [reply design/new-prepost-backuprestore-plugin-hooks.md:112]: /kind changelog-not-required

**Review by dsu-igeek — APPROVED** [review]
LGTM, thanks!

File: design/new-prepost-backuprestore-plugin-hooks.md
Comment (jglick) [reply design/new-prepost-backuprestore-plugin-hooks.md:106]: Agreed. I thought there was already an issue about this topic, but if so I cannot find it now due to https://github.com/github/feedback/discussions/6403.

**Pull request merged**
